File: lib/services/autocompletion_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:openai_dart/openai_dart.dart';
import 'config_service.dart';
import '../core/tool.dart';
import 'sse_stream_client.dart';

/// Configuration keys for the autocompletion service (global scope)
class AutocompletionConfig {
  static const String baseUrl = 'llm.base_url';
  static const String apiKey = 'llm.api_key';
  static const String model = 'llm.model';
  static const String maxTokens = 'llm.max_tokens';
  static const String temperature = 'llm.temperature';
  static const String summaryMaxTokens = 'llm.summary_max_tokens';

  // Default values
  static const String defaultBaseUrl = 'https://api.openai.com/v1';
  static const String defaultModel = 'gpt-4o-mini';
  static const String defaultMaxTokens = '1024';
  static const String defaultTemperature = '0.7';
  static const String defaultSummaryMaxTokens = '4096';
}

/// Message role in a chat conversation
enum MessageRole {
  system,
  user,
  assistant,
  tool,
}

/// Represents a tool call made by the assistant
class ChatToolCall {
  final String id;
  final String name;
  final Map<String, dynamic> arguments;

  const ChatToolCall({
    required this.id,
    required this.name,
    required this.arguments,
  });
}

/// A chat message with role and content
class ChatMessage {
  final MessageRole role;
  final String content;
  final List<ChatToolCall>? toolCalls;
  final String? toolCallId;

  const ChatMessage({
    required this.role,
    required this.content,
    this.toolCalls,
    this.toolCallId,
  });

  ChatCompletionMessage toOpenAI() {
    switch (role) {
      case MessageRole.system:
        return ChatCompletionMessage.system(content: content);
      case MessageRole.user:
        return ChatCompletionMessage.user(
          content: ChatCompletionUserMessageContent.string(content),
        );
      case MessageRole.assistant:
        if (toolCalls != null && toolCalls!.isNotEmpty) {
          return ChatCompletionMessage.assistant(
            content: content.isEmpty ? null : content,
            toolCalls: toolCalls!
                .map((tc) => ChatCompletionMessageToolCall(
                      id: tc.id,
                      type: ChatCompletionMessageToolCallType.function,
                      function: ChatCompletionMessageFunctionCall(
                        name: tc.name,
                        arguments: jsonEncode(tc.arguments),
                      ),
                    ))
                .toList(),
          );
        }
        return ChatCompletionMessage.assistant(content: content);
      case MessageRole.tool:
        return ChatCompletionMessage.tool(
          toolCallId: toolCallId ?? '',
          content: content,
        );
    }
  }
}

/// Response from a completion request
class CompletionResponse {
  final String content;
  final String? finishReason;
  final int? promptTokens;
  final int? completionTokens;

  const CompletionResponse({
    required this.content,
    this.finishReason,
    this.promptTokens,
    this.completionTokens,
  });
}

/// Events emitted during streaming completion with tools
sealed class ChatStreamEvent {}

/// A chunk of content from the assistant
class ContentChunk extends ChatStreamEvent {
  final String content;
  ContentChunk(this.content);
}

/// A tool call requested by the assistant
class ToolCallEvent extends ChatStreamEvent {
  final String id;
  final String name;
  final Map<String, dynamic> arguments;
  ToolCallEvent(this.id, this.name, this.arguments);
}

/// Indicates the completion is done
class CompletionDone extends ChatStreamEvent {}

/// Helper class to build tool calls from streaming chunks
class _ToolCallBuilder {
  String? id;
  String? name;
  final StringBuffer argumentsBuffer = StringBuffer();
}

/// Service for generating text completions using OpenAI-compatible APIs.
/// Supports OpenAI, vLLM, Ollama, and other compatible endpoints.
class AutocompletionService {
  static AutocompletionService? _instance;
  static AutocompletionService get instance =>
      _instance ??= AutocompletionService._();

  AutocompletionService._();

  OpenAIClient? _client;
  String? _currentBaseUrl;
  String? _currentApiKey;

  /// Initialize global config defaults. Call this during app startup.
  static void initializeDefaults() {
    final config = ConfigService.instance;
    config.setDefault(AutocompletionConfig.baseUrl, AutocompletionConfig.defaultBaseUrl);
    config.setDefault(AutocompletionConfig.model, AutocompletionConfig.defaultModel);
    config.setDefault(AutocompletionConfig.maxTokens, AutocompletionConfig.defaultMaxTokens);
    config.setDefault(AutocompletionConfig.temperature, AutocompletionConfig.defaultTemperature);
    config.setDefault(AutocompletionConfig.summaryMaxTokens, AutocompletionConfig.defaultSummaryMaxTokens);
    // API key has no default - must be configured by user
  }

  /// Get or create the OpenAI client with current configuration
  OpenAIClient _getClient() {
    final config = ConfigService.instance;
    final baseUrl = config.get(AutocompletionConfig.baseUrl) ??
        AutocompletionConfig.defaultBaseUrl;
    final apiKey = config.get(AutocompletionConfig.apiKey);

    // Recreate client if config changed
    if (_client == null ||
        _currentBaseUrl != baseUrl ||
        _currentApiKey != apiKey) {
      _client = OpenAIClient(
        apiKey: apiKey ?? '',
        baseUrl: baseUrl,
      );
      _currentBaseUrl = baseUrl;
      _currentApiKey = apiKey;
    }

    return _client!;
  }

  /// Check if the service is configured (has API key)
  bool get isConfigured {
    final apiKey = ConfigService.instance.get(AutocompletionConfig.apiKey);
    return apiKey != null && apiKey.isNotEmpty;
  }

  /// Get the current model
  String get currentModel {
    return ConfigService.instance.get(AutocompletionConfig.model) ??
        AutocompletionConfig.defaultModel;
  }

  /// Get the current base URL
  String get currentBaseUrl {
    return ConfigService.instance.get(AutocompletionConfig.baseUrl) ??
        AutocompletionConfig.defaultBaseUrl;
  }

  /// Generate a completion from a list of messages
  Future<CompletionResponse> complete(
    List<ChatMessage> messages, {
    String? model,
    int? maxTokens,
    double? temperature,
  }) async {
    if (!isConfigured) {
      throw StateError(
        'AutocompletionService not configured. Set ${AutocompletionConfig.apiKey} in config.',
      );
    }

    final config = ConfigService.instance;
    final effectiveModel =
        model ?? config.get(AutocompletionConfig.model) ?? AutocompletionConfig.defaultModel;
    final effectiveMaxTokens = maxTokens ??
        int.tryParse(config.get(AutocompletionConfig.maxTokens) ?? '') ??
        int.parse(AutocompletionConfig.defaultMaxTokens);
    final effectiveTemperature = temperature ??
        double.tryParse(config.get(AutocompletionConfig.temperature) ?? '') ??
        double.parse(AutocompletionConfig.defaultTemperature);

    final client = _getClient();

    final response = await client.createChatCompletion(
      request: CreateChatCompletionRequest(
        model: ChatCompletionModel.modelId(effectiveModel),
        messages: messages.map((m) => m.toOpenAI()).toList(),
        maxCompletionTokens: effectiveMaxTokens,
        temperature: effectiveTemperature,
      ),
    );

    final choice = response.choices.firstOrNull;

    // For reasoning models: content is the final answer, reasoningContent is the thinking
    // Some models (vLLM reasoning) only return reasoningContent and never content
    final content = choice?.message.content;
    final reasoning = choice?.message.reasoningContent;

    String finalContent;
    if (content != null && content.isNotEmpty) {
      // Normal case: model returned final answer in content
      finalContent = content;
    } else if (reasoning != null && reasoning.isNotEmpty) {
      // Reasoning model case: need to extract answer from reasoning
      finalContent = reasoning;
    } else {
      finalContent = '';
    }

    return CompletionResponse(
      content: finalContent,
      finishReason: choice?.finishReason?.name,
      promptTokens: response.usage?.promptTokens,
      completionTokens: response.usage?.completionTokens,
    );
  }

  /// Generate a completion with streaming response
  Stream<String> completeStream(
    List<ChatMessage> messages, {
    String? model,
    int? maxTokens,
    double? temperature,
  }) async* {
    if (!isConfigured) {
      throw StateError(
        'AutocompletionService not configured. Set ${AutocompletionConfig.apiKey} in config.',
      );
    }

    final config = ConfigService.instance;
    final effectiveModel =
        model ?? config.get(AutocompletionConfig.model) ?? AutocompletionConfig.defaultModel;
    final effectiveMaxTokens = maxTokens ??
        int.tryParse(config.get(AutocompletionConfig.maxTokens) ?? '') ??
        int.parse(AutocompletionConfig.defaultMaxTokens);
    final effectiveTemperature = temperature ??
        double.tryParse(config.get(AutocompletionConfig.temperature) ?? '') ??
        double.parse(AutocompletionConfig.defaultTemperature);

    final client = _getClient();

    final stream = client.createChatCompletionStream(
      request: CreateChatCompletionRequest(
        model: ChatCompletionModel.modelId(effectiveModel),
        messages: messages.map((m) => m.toOpenAI()).toList(),
        maxCompletionTokens: effectiveMaxTokens,
        temperature: effectiveTemperature,
      ),
    );

    await for (final chunk in stream) {
      final choice = chunk.choices?.firstOrNull;
      if (choice == null) continue;

      // For reasoning models: prefer content, but fallback to reasoningContent if content is never sent
      // Some vLLM reasoning models ONLY stream reasoningContent in streaming mode
      final delta = choice.delta?.content ?? choice.delta?.reasoningContent;

      if (delta != null && delta.isNotEmpty) {
        yield delta;
      }
    }
  }

  /// Generate a streaming completion with proper separation of reasoning and content.
  ///
  /// Uses raw SSE parsing to correctly handle vLLM reasoning models that send
  /// chain-of-thought in `reasoning_content` and final answer in `content`.
  ///
  /// Set [contentOnly] to true to only yield main content (skip reasoning).
  /// This is useful for structured outputs like definitions where you only want
  /// the final answer without chain-of-thought.
  Stream<StreamChunk> completeStreamStructured(
    List<ChatMessage> messages, {
    String? model,
    int? maxTokens,
    double? temperature,
    bool contentOnly = false,
  }) async* {
    if (!isConfigured) {
      throw StateError(
        'AutocompletionService not configured. Set ${AutocompletionConfig.apiKey} in config.',
      );
    }

    final config = ConfigService.instance;
    final effectiveModel =
        model ?? config.get(AutocompletionConfig.model) ?? AutocompletionConfig.defaultModel;
    final effectiveMaxTokens = maxTokens ??
        int.tryParse(config.get(AutocompletionConfig.maxTokens) ?? '') ??
        int.parse(AutocompletionConfig.defaultMaxTokens);
    final effectiveTemperature = temperature ??
        double.tryParse(config.get(AutocompletionConfig.temperature) ?? '') ??
        double.parse(AutocompletionConfig.defaultTemperature);
    final baseUrl = config.get(AutocompletionConfig.baseUrl) ??
        AutocompletionConfig.defaultBaseUrl;
    final apiKey = config.get(AutocompletionConfig.apiKey) ?? '';

    final sseClient = SseStreamClient(
      baseUrl: baseUrl,
      apiKey: apiKey,
    );

    final openAIMessages = messages.map((m) => {
      'role': m.role.name,
      'content': m.content,
    }).toList();

    await for (final chunk in sseClient.streamChatCompletion(
      model: effectiveModel,
      messages: openAIMessages,
      temperature: effectiveTemperature,
      maxTokens: effectiveMaxTokens,
    )) {
      if (contentOnly) {
        // Only yield chunks with main content (skip reasoning)
        if (chunk.hasContent || chunk.isDone) {
          yield chunk;
        }
      } else {
        yield chunk;
      }
    }
  }

  /// Generate a completion with tools support (streaming)
  /// Yields ChatStreamEvent objects for content chunks and tool calls
  Stream<ChatStreamEvent> completeWithTools(
    List<ChatMessage> messages, {
    List<Tool>? tools,
    String? model,
    int? maxTokens,
    double? temperature,
  }) async* {
    if (!isConfigured) {
      throw StateError(
        'AutocompletionService not configured. Set ${AutocompletionConfig.apiKey} in config.',
      );
    }

    final config = ConfigService.instance;
    final effectiveModel =
        model ?? config.get(AutocompletionConfig.model) ?? AutocompletionConfig.defaultModel;
    final effectiveMaxTokens = maxTokens ??
        int.tryParse(config.get(AutocompletionConfig.maxTokens) ?? '') ??
        int.parse(AutocompletionConfig.defaultMaxTokens);
    final effectiveTemperature = temperature ??
        double.tryParse(config.get(AutocompletionConfig.temperature) ?? '') ??
        double.parse(AutocompletionConfig.defaultTemperature);

    final client = _getClient();

    // Build tools in OpenAI format
    final openAITools = tools?.map((tool) => ChatCompletionTool(
          type: ChatCompletionToolType.function,
          function: FunctionObject(
            name: tool.name,
            description: tool.description,
            parameters: tool.parameters,
          ),
        )).toList();

    final stream = client.createChatCompletionStream(
      request: CreateChatCompletionRequest(
        model: ChatCompletionModel.modelId(effectiveModel),
        messages: messages.map((m) => m.toOpenAI()).toList(),
        maxCompletionTokens: effectiveMaxTokens,
        temperature: effectiveTemperature,
        tools: openAITools,
      ),
    );

    // Track tool calls being built up from streaming chunks
    final toolCallsInProgress = <String, _ToolCallBuilder>{};

    await for (final chunk in stream) {
      final choice = chunk.choices?.firstOrNull;
      if (choice == null) continue;

      // Only use content, not reasoningContent (reasoning is sent before the actual answer)
      final contentDelta = choice.delta?.content;
      if (contentDelta != null && contentDelta.isNotEmpty) {
        yield ContentChunk(contentDelta);
      }

      // Handle tool call deltas
      final toolCallDeltas = choice.delta?.toolCalls;
      if (toolCallDeltas != null) {
        for (final toolCallDelta in toolCallDeltas) {
          final index = toolCallDelta.index;
          final indexKey = index.toString();

          // Initialize builder if this is a new tool call
          if (!toolCallsInProgress.containsKey(indexKey)) {
            toolCallsInProgress[indexKey] = _ToolCallBuilder();
          }

          final builder = toolCallsInProgress[indexKey]!;

          // Accumulate data from delta
          if (toolCallDelta.id != null) {
            builder.id = toolCallDelta.id!;
          }
          if (toolCallDelta.function?.name != null) {
            builder.name = toolCallDelta.function!.name!;
          }
          if (toolCallDelta.function?.arguments != null) {
            builder.argumentsBuffer.write(toolCallDelta.function!.arguments!);
          }
        }
      }

      // Check if stream is done
      final finishReason = choice.finishReason;
      if (finishReason != null) {
        // If finished with tool_calls, emit the tool call events
        if (finishReason == ChatCompletionFinishReason.toolCalls) {
          for (final builder in toolCallsInProgress.values) {
            if (builder.id != null && builder.name != null) {
              Map<String, dynamic> args = {};
              try {
                final argsString = builder.argumentsBuffer.toString();
                if (argsString.isNotEmpty) {
                  args = jsonDecode(argsString) as Map<String, dynamic>;
                }
              } catch (_) {
                // Skip malformed arguments
              }
              yield ToolCallEvent(builder.id!, builder.name!, args);
            }
          }
        }

        yield CompletionDone();
      }
    }
  }

  /// Simple completion with a single prompt (no conversation context)
  Future<String> prompt(
    String prompt, {
    String? systemPrompt,
    String? model,
    int? maxTokens,
    double? temperature,
  }) async {
    final messages = <ChatMessage>[
      if (systemPrompt != null)
        ChatMessage(role: MessageRole.system, content: systemPrompt),
      ChatMessage(role: MessageRole.user, content: prompt),
    ];

    final response = await complete(
      messages,
      model: model,
      maxTokens: maxTokens,
      temperature: temperature,
    );

    return response.content;
  }

  /// Simple streaming completion with a single prompt
  Stream<String> promptStream(
    String prompt, {
    String? systemPrompt,
    String? model,
    int? maxTokens,
    double? temperature,
  }) {
    final messages = <ChatMessage>[
      if (systemPrompt != null)
        ChatMessage(role: MessageRole.system, content: systemPrompt),
      ChatMessage(role: MessageRole.user, content: prompt),
    ];

    return completeStream(
      messages,
      model: model,
      maxTokens: maxTokens,
      temperature: temperature,
    );
  }

  /// Simple streaming completion with structured output (content only, no reasoning)
  ///
  /// Uses raw SSE streaming to properly separate reasoning from content.
  /// Yields main content immediately. If model only sends reasoning, yields that instead.
  /// This is ideal for structured outputs like titles, definitions, etc.
  Stream<String> promptStreamContentOnly(
    String prompt, {
    String? systemPrompt,
    String? model,
    int? maxTokens,
    double? temperature,
  }) async* {
    final messages = <ChatMessage>[
      if (systemPrompt != null)
        ChatMessage(role: MessageRole.system, content: systemPrompt),
      ChatMessage(role: MessageRole.user, content: prompt),
    ];

    bool hasYieldedContent = false;
    final reasoningBuffer = StringBuffer();

    await for (final chunk in completeStreamStructured(
      messages,
      model: model,
      maxTokens: maxTokens,
      temperature: temperature,
      contentOnly: false,  // Get both reasoning and content
    )) {
      if (chunk.hasContent) {
        hasYieldedContent = true;
        yield chunk.content!;
      } else if (chunk.hasReasoning) {
        // Buffer reasoning — only yield if no content is ever produced
        reasoningBuffer.write(chunk.reasoningContent!);
      }
    }

    // Only fall back to reasoning if the model never sent any content
    if (!hasYieldedContent && reasoningBuffer.isNotEmpty) {
      yield reasoningBuffer.toString();
    }
  }

  /// Update API configuration
  Future<void> configure({
    String? baseUrl,
    String? apiKey,
    String? model,
    int? maxTokens,
    double? temperature,
  }) async {
    final config = ConfigService.instance;

    if (baseUrl != null) {
      await config.set(AutocompletionConfig.baseUrl, baseUrl);
    }
    if (apiKey != null) {
      await config.set(AutocompletionConfig.apiKey, apiKey);
    }
    if (model != null) {
      await config.set(AutocompletionConfig.model, model);
    }
    if (maxTokens != null) {
      await config.set(AutocompletionConfig.maxTokens, maxTokens.toString());
    }
    if (temperature != null) {
      await config.set(AutocompletionConfig.temperature, temperature.toString());
    }

    // Reset client to pick up new config
    _client = null;
  }

  /// Reset instance for testing
  static void resetInstance() {
    _instance?._client = null;
    _instance = null;
  }
}

File: playground/Playground/Playground/Sources/Services/AutocompletionService.swift
import Foundation
import Alamofire

/// OpenAI-compatible LLM API client
/// Supports both streaming and non-streaming completions
class AutocompletionService {
    static let shared = AutocompletionService()
    private let config = ConfigService.shared

    private init() {}

    /// Check if the service is properly configured
    var isConfigured: Bool {
        let baseURL = config.getString(key: "llm.base_url")
        return !baseURL.isEmpty
    }

    // MARK: - Chat Completion

    struct ChatMessage: Codable, Sendable {
        let role: String
        let content: String
    }

    struct ChatCompletionRequest: Codable, Sendable {
        let model: String
        let messages: [ChatMessage]
        let temperature: Double?
        let maxTokens: Int?
        let stream: Bool
        let tools: [[String: Any]]?

        enum CodingKeys: String, CodingKey {
            case model
            case messages
            case temperature
            case maxTokens = "max_tokens"
            case stream
            case tools
        }

        init(model: String, messages: [ChatMessage], temperature: Double?, maxTokens: Int?, stream: Bool, tools: [[String: Any]]? = nil) {
            self.model = model
            self.messages = messages
            self.temperature = temperature
            self.maxTokens = maxTokens
            self.stream = stream
            self.tools = tools
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            model = try container.decode(String.self, forKey: .model)
            messages = try container.decode([ChatMessage].self, forKey: .messages)
            temperature = try container.decodeIfPresent(Double.self, forKey: .temperature)
            maxTokens = try container.decodeIfPresent(Int.self, forKey: .maxTokens)
            stream = try container.decode(Bool.self, forKey: .stream)
            tools = nil // Tools are not decoded from JSON
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.container(keyedBy: CodingKeys.self)
            try container.encode(model, forKey: .model)
            try container.encode(messages, forKey: .messages)
            try container.encodeIfPresent(temperature, forKey: .temperature)
            try container.encodeIfPresent(maxTokens, forKey: .maxTokens)
            try container.encode(stream, forKey: .stream)

            // Manually encode tools as JSON
            if let tools = tools {
                // We need to manually serialize this since it's [String: Any]
                // This will be handled by custom JSON encoding
            }
        }

        func toJSONData() throws -> Data {
            var dict: [String: Any] = [
                "model": model,
                "messages": messages.map { ["role": $0.role, "content": $0.content] },
                "stream": stream
            ]

            if let temperature = temperature {
                dict["temperature"] = temperature
            }
            if let maxTokens = maxTokens {
                dict["max_tokens"] = maxTokens
            }
            if let tools = tools, !tools.isEmpty {
                dict["tools"] = tools
            }

            return try JSONSerialization.data(withJSONObject: dict)
        }
    }

    struct ChatCompletionResponse: Codable, Sendable {
        let id: String?
        let choices: [Choice]

        struct Choice: Codable, Sendable {
            let message: ChatMessage
            let finishReason: String?

            enum CodingKeys: String, CodingKey {
                case message
                case finishReason = "finish_reason"
            }
        }
    }

    struct ChatCompletionChunk: Codable, Sendable {
        let id: String?
        let choices: [Choice]

        struct Choice: Codable, Sendable {
            let delta: Delta
            let finishReason: String?

            enum CodingKeys: String, CodingKey {
                case delta
                case finishReason = "finish_reason"
            }
        }

        struct Delta: Codable, Sendable {
            let role: String?
            let content: String?
            let toolCalls: [ToolCallDelta]?

            enum CodingKeys: String, CodingKey {
                case role
                case content
                case toolCalls = "tool_calls"
            }
        }

        struct ToolCallDelta: Codable, Sendable {
            let index: Int
            let id: String?
            let type: String?
            let function: FunctionDelta?

            struct FunctionDelta: Codable, Sendable {
                let name: String?
                let arguments: String?
            }
        }
    }

    /// Stream events for tool calling
    enum ChatStreamEvent {
        case contentChunk(String)
        case toolCall(id: String, name: String, arguments: [String: Any])
        case done
    }

    /// Non-streaming chat completion
    func complete(messages: [ChatMessage],
                  model: String? = nil,
                  temperature: Double? = nil,
                  maxTokens: Int? = nil) async -> Result<String, Error> {
        return await Result.catching {
            let baseURL = self.config.getString(key: "llm.base_url")
            let apiKey = self.config.getString(key: "llm.api_key")
            let modelName = model ?? self.config.getString(key: "llm.model", default: "gpt-4o-mini")

            let request = ChatCompletionRequest(
                model: modelName,
                messages: messages,
                temperature: temperature ?? self.config.getDouble(key: "llm.temperature", default: 0.7),
                maxTokens: maxTokens ?? self.config.getInt(key: "llm.max_tokens", default: 1024),
                stream: false,
                tools: nil
            )

            // Encode request manually to avoid Sendable conformance issues
            let requestBody = try request.toJSONData()

            var headers = HTTPHeaders()
            if !apiKey.isEmpty {
                headers.add(.authorization(bearerToken: apiKey))
            }
            headers.add(.contentType("application/json"))

            let responseData = try await AF.request(
                "\(baseURL)/chat/completions",
                method: .post,
                headers: headers
            ) { urlRequest in
                urlRequest.httpBody = requestBody
            }
            .serializingData()
            .value

            // Decode response manually to avoid Sendable conformance issues
            let response = try JSONDecoder().decode(ChatCompletionResponse.self, from: responseData)
            return response.choices.first?.message.content ?? ""
        }
    }

    /// Streaming chat completion
    func completeStream(messages: [ChatMessage],
                       model: String? = nil,
                       temperature: Double? = nil,
                       maxTokens: Int? = nil) -> AsyncThrowingStream<String, Error> {
        let baseURL = config.getString(key: "llm.base_url")
        let apiKey = config.getString(key: "llm.api_key")
        let modelName = model ?? config.getString(key: "llm.model", default: "gpt-4o-mini")

        let request = ChatCompletionRequest(
            model: modelName,
            messages: messages,
            temperature: temperature ?? config.getDouble(key: "llm.temperature", default: 0.7),
            maxTokens: maxTokens ?? config.getInt(key: "llm.max_tokens", default: 1024),
            stream: true,
            tools: nil
        )

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    var headers = HTTPHeaders()
                    if !apiKey.isEmpty {
                        headers.add(.authorization(bearerToken: apiKey))
                    }
                    headers.add(.contentType("application/json"))

                    let requestBody = try request.toJSONData()

                    var urlRequest = URLRequest(url: URL(string: "\(baseURL)/chat/completions")!)
                    urlRequest.httpMethod = "POST"
                    urlRequest.httpBody = requestBody
                    for header in headers.dictionary {
                        urlRequest.setValue(header.value, forHTTPHeaderField: header.key)
                    }

                    let (bytes, response) = try await URLSession.shared.bytes(for: urlRequest)

                    guard let httpResponse = response as? HTTPURLResponse else {
                        print("❌ Response is not HTTPURLResponse")
                        throw AutocompletionError.invalidResponse
                    }

                    guard httpResponse.statusCode == 200 else {
                        print("❌ HTTP Status: \(httpResponse.statusCode)")
                        print("❌ URL: \(baseURL)/chat/completions")
                        print("❌ Model: \(modelName)")
                        throw AutocompletionError.httpError(statusCode: httpResponse.statusCode)
                    }

                    for try await line in bytes.lines {
                        guard line.hasPrefix("data: ") else { continue }
                        let data = String(line.dropFirst(6))

                        if data == "[DONE]" {
                            continuation.finish()
                            return
                        }

                        guard let jsonData = data.data(using: .utf8) else { continue }
                        let chunk = try JSONDecoder().decode(ChatCompletionChunk.self, from: jsonData)

                        if let content = chunk.choices.first?.delta.content {
                            continuation.yield(content)
                        }

                        if chunk.choices.first?.finishReason != nil {
                            continuation.finish()
                            return
                        }
                    }

                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    /// Streaming chat completion with tools support
    /// Yields ChatStreamEvent objects for content chunks and tool calls
    func completeWithTools(messages: [ChatMessage],
                          tools: [[String: Any]]?,
                          model: String? = nil,
                          temperature: Double? = nil,
                          maxTokens: Int? = nil) -> AsyncThrowingStream<ChatStreamEvent, Error> {
        let baseURL = config.getString(key: "llm.base_url")
        let apiKey = config.getString(key: "llm.api_key")
        let modelName = model ?? config.getString(key: "llm.model", default: "gpt-4o-mini")

        let request = ChatCompletionRequest(
            model: modelName,
            messages: messages,
            temperature: temperature ?? config.getDouble(key: "llm.temperature", default: 0.7),
            maxTokens: maxTokens ?? config.getInt(key: "llm.max_tokens", default: 1024),
            stream: true,
            tools: tools
        )

        return AsyncThrowingStream { continuation in
            Task {
                do {
                    var headers = HTTPHeaders()
                    if !apiKey.isEmpty {
                        headers.add(.authorization(bearerToken: apiKey))
                    }
                    headers.add(.contentType("application/json"))

                    let requestBody = try request.toJSONData()

                    var urlRequest = URLRequest(url: URL(string: "\(baseURL)/chat/completions")!)
                    urlRequest.httpMethod = "POST"
                    urlRequest.httpBody = requestBody
                    for header in headers.dictionary {
                        urlRequest.setValue(header.value, forHTTPHeaderField: header.key)
                    }

                    let (bytes, response) = try await URLSession.shared.bytes(for: urlRequest)

                    guard let httpResponse = response as? HTTPURLResponse else {
                        print("❌ Response is not HTTPURLResponse")
                        throw AutocompletionError.invalidResponse
                    }

                    guard httpResponse.statusCode == 200 else {
                        print("❌ HTTP Status: \(httpResponse.statusCode)")
                        throw AutocompletionError.httpError(statusCode: httpResponse.statusCode)
                    }

                    // Track tool calls being built up from streaming chunks
                    var toolCallsInProgress: [Int: (id: String?, name: String?, arguments: String)] = [:]

                    for try await line in bytes.lines {
                        guard line.hasPrefix("data: ") else { continue }
                        let data = String(line.dropFirst(6))

                        if data == "[DONE]" {
                            continuation.yield(.done)
                            continuation.finish()
                            return
                        }

                        guard let jsonData = data.data(using: .utf8) else { continue }
                        let chunk = try JSONDecoder().decode(ChatCompletionChunk.self, from: jsonData)

                        guard let choice = chunk.choices.first else { continue }

                        // Handle content
                        if let content = choice.delta.content {
                            continuation.yield(.contentChunk(content))
                        }

                        // Handle tool call deltas
                        if let toolCallDeltas = choice.delta.toolCalls {
                            for toolCallDelta in toolCallDeltas {
                                let index = toolCallDelta.index

                                // Initialize builder if new tool call
                                if toolCallsInProgress[index] == nil {
                                    toolCallsInProgress[index] = (nil, nil, "")
                                }

                                var current = toolCallsInProgress[index]!

                                // Accumulate data from delta
                                if let id = toolCallDelta.id {
                                    current.id = id
                                }
                                if let name = toolCallDelta.function?.name {
                                    current.name = name
                                }
                                if let args = toolCallDelta.function?.arguments {
                                    current.arguments += args
                                }

                                toolCallsInProgress[index] = current
                            }
                        }

                        // Check if stream is done
                        if let finishReason = choice.finishReason {
                            // If finished with tool_calls, emit the tool call events
                            if finishReason == "tool_calls" {
                                for (_, toolCall) in toolCallsInProgress {
                                    if let id = toolCall.id, let name = toolCall.name {
                                        var args: [String: Any] = [:]
                                        if !toolCall.arguments.isEmpty,
                                           let argsData = toolCall.arguments.data(using: .utf8),
                                           let json = try? JSONSerialization.jsonObject(with: argsData) as? [String: Any] {
                                            args = json
                                        }
                                        continuation.yield(.toolCall(id: id, name: name, arguments: args))
                                    }
                                }
                            }

                            continuation.yield(.done)
                            continuation.finish()
                            return
                        }
                    }

                    continuation.yield(.done)
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - Convenience Methods

    /// Simple prompt-based completion (user message)
    func prompt(_ userMessage: String,
                systemPrompt: String? = nil,
                temperature: Double? = nil,
                maxTokens: Int? = nil) async -> Result<String, Error> {
        var messages: [ChatMessage] = []

        if let systemPrompt = systemPrompt {
            messages.append(ChatMessage(role: "system", content: systemPrompt))
        }

        messages.append(ChatMessage(role: "user", content: userMessage))

        return await complete(
            messages: messages,
            temperature: temperature,
            maxTokens: maxTokens
        )
    }

    /// Simple streaming prompt
    func promptStream(_ userMessage: String,
                     systemPrompt: String? = nil,
                     temperature: Double? = nil,
                     maxTokens: Int? = nil) -> AsyncThrowingStream<String, Error> {
        var messages: [ChatMessage] = []

        if let systemPrompt = systemPrompt {
            messages.append(ChatMessage(role: "system", content: systemPrompt))
        }

        messages.append(ChatMessage(role: "user", content: userMessage))

        return completeStream(
            messages: messages,
            temperature: temperature,
            maxTokens: maxTokens
        )
    }
}

enum AutocompletionError: Error {
    case invalidResponse
    case httpError(statusCode: Int)
    case missingContent
    case networkError(Error)

    var localizedDescription: String {
        switch self {
        case .invalidResponse:
            return "Invalid response from server"
        case .httpError(let statusCode):
            return "HTTP error: \(statusCode)"
        case .missingContent:
            return "Missing content in response"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }
}
